autoload/phpactor/project/projectRepository.vim

```diff
@@ -0,0 +1,47 @@
+function! phpactor#project#projectRepository#create() abort
+  return {
+        \ 'projects': {},
+        \ 'isEmpty': function('s:isEmpty'),
+        \ 'addProject': function('s:addProject'),
+        \ 'findProjectContainingFile': function('s:findProjectContainingFile'),
+        \ 'listProjects': function('s:listProjects'),
+        \ 'hasProjectWithPrimaryRoot': function('s:hasProjectWithPrimaryRoot'),
+        \ 'hasProject': function('s:hasProject')
+        \ }
+endfunction
+
+function! s:listProjects() dict abort
+  return copy(self.projects)
+endfunction
+
+function! s:findProjectContainingFile(file) dict abort
+    for l:rootPath in keys(self.projects)
+      let l:project = self.projects[l:rootPath]
+      if l:project.containsFile(a:file)
+        return l:project
+      endif
+    endfor
+
+    return v:null
+endfunction
+
+function! s:addProject(project) dict abort
+  " todo check type
+  if v:null isnot get(self.projects, a:project.getPrimaryRootPath(), v:null)
+    throw printf('Project "%s" already exists', a:project.getPrimaryRootPath())
+  endif
+
+  let self.projects[a:project.getPrimaryRootPath()] = a:project
+endfunction
+
+function! s:hasProject(project) dict abort
+  return get(self.projects, a:project.getPrimaryRootPath(), v:false) isnot v:false
```

Comment:
Here you don't have to check the all structure but at least that the function exist and is a function.
But in order to avoid repeating the same checks and to keep the knowledge at the same place I encourage your to create a function `phpactor#project#project#isValid(project)` that will do all the necessary tests.